mmengine/utils/progressbar_rich.py

```diff
@@ -0,0 +1,365 @@
+# Copyright (c) OpenMMLab. All rights reserved.
+import threading
+import time
+from collections.abc import Iterable
+from multiprocessing import Pool
+from typing import List
+
+from rich.console import Console
+from rich.progress import Progress
+
+from .timer import Timer
+
+
+class RichProgressBar:
+    """use rich to enhance progress bar."""
+
+    def __init__(self):
+        self.bar = Progress()
+        self.console = Console()
+        self.tasks = []
+        self.colors = []
+
+        self.bar.start()
+
+    def __del__(self):
+        self.bar.stop()
+
+    def write(self, msg, color='blue'):
+        self.console.print(msg, style=color)
+
+    def add_multi_task(self, total, color='blue', description='Process...'):
+        assert total > 0
+        self.colors.append(color)
+        self.tasks.append(
+            self.bar.add_task(f'[{color}]{description}', total=total))
+
+    def add_single_task(self, total, color='blue', description='Process...'):
+        self.colors.append(color)
+        if total > 0:
+            self.tasks.append(
+                self.bar.add_task(f'[{color}]{description}', total=total))
+            self.infinite = False
+        else:
+            self.write('completed: 0, elapsed: 0s', self.colors[0])
+            self.timer = Timer()
+            self.infinite = True
+            self.completed = 0
+            self.tasks.append(0)
+
+    def is_task_finish(self, task_id):
+        completed = self.bar.tasks[task_id].completed
+        total = self.bar.tasks[task_id].total
+        return completed == total
+
+    def finished(self):
+        return self.bar.finished
+
+    def update(self, task_id=0, advance=1):
+        assert advance > 0
+
+        if len(self.tasks) == 1:
+            assert task_id == 0
+            if not self.infinite:
+                self.bar.advance(self.tasks[task_id], advance=advance)
+
+                if self.is_task_finish(task_id):
+                    self.bar.stop()
+            else:
+                self.completed += advance
+                elapsed = self.timer.since_start()
+                if elapsed > 0:
+                    fps = self.completed / elapsed
+                else:
+                    fps = float('inf')
+                self.write(
+                    f'completed: {self.completed},'
+                    f' elapsed: {int(elapsed + 0.5)}s,'
+                    f' {fps:.1f} tasks/s', self.colors[task_id])
+        else:
+            assert task_id >= 0 and task_id < len(self.tasks)
+            self.bar.update(self.tasks[task_id], advance=advance)
```

Comment:
Does it seem we don't consider `finite` and `infinite` tasks exist at the same time? If so, we could raise an error when users calls `add_single_track`.

Reply:
when user add a single track, the code won`t get this path